spatialos-sdk/src/worker/schema.rs

```diff
@@ -0,0 +1,662 @@
+    Schema_AddEnum,
+    Schema_AddEnumList,
+    Schema_GetEnumList,
+);
+
+impl Sealed for EntityId {}
+
+impl SchemaField for EntityId {
+    type RustType = Self;
+
+    fn add_field(object: &mut SchemaObject, field: FieldId, value: &Self::RustType) {
+        unsafe {
+            Schema_AddEntityId(object.internal, field, value.id);
+        }
+    }
+
+    fn get_field(object: &SchemaObject, field: FieldId) -> Self::RustType {
+        let id = unsafe { Schema_GetEntityId(object.internal, field) };
+        Self { id }
+    }
+}
+
+impl IndexedField for EntityId {
+    fn field_count(object: &SchemaObject, field: FieldId) -> u32 {
+        unsafe { Schema_GetEntityIdCount(object.internal, field) }
+    }
+
+    fn index_field(object: &SchemaObject, field: FieldId, index: u32) -> Self::RustType {
+        let id = unsafe { Schema_IndexEntityId(object.internal, field, index) };
+        Self { id }
+    }
+}
+
+// NOTE: It's safe to treat the `EntityId` array as an `i64` array because
+// `EntityId` is marked `repr(transparent)`, and so is identical to an `i64` for
+// the purpose of FFI.
+impl ArrayField for EntityId {
+    fn add_field_list(object: &mut SchemaObject, field: FieldId, data: &[Self::RustType]) {
+        unsafe {
+            Schema_AddEntityIdList(
+                object.internal,
+                field,
+                data.as_ptr() as *const _,
+                data.len() as u32,
+            );
+        }
+    }
+
+    fn get_field_list(object: &SchemaObject, field: FieldId, data: &mut Vec<Self::RustType>) {
+        let count = Self::field_count(object, field) as usize;
+
+        // Ensure that there is enough capacity for the elements in the schema field.
+        if data.capacity() < count {
+            data.reserve(count - data.capacity());
+        }
+
+        // Replace the contents of `data` with the list of values in the schema field.
+        unsafe {
+            data.set_len(count);
+            Schema_GetEntityIdList(object.internal, field, data.as_mut_ptr() as *mut _);
+        }
+    }
+}
+
+impl Sealed for bool {}
+
+impl SchemaField for bool {
+    type RustType = Self;
+
+    fn add_field(object: &mut SchemaObject, field: FieldId, value: &Self::RustType) {
+        unsafe {
+            Schema_AddBool(object.internal, field, *value as u8);
+        }
+    }
+
+    fn get_field(object: &SchemaObject, field: FieldId) -> Self::RustType {
+        let raw = unsafe { Schema_GetBool(object.internal, field) };
+        raw != 0
+    }
+}
+
+impl IndexedField for bool {
+    fn field_count(object: &SchemaObject, field: FieldId) -> u32 {
+        unsafe { Schema_GetBoolCount(object.internal, field) }
+    }
+
+    fn index_field(object: &SchemaObject, field: FieldId, index: u32) -> Self::RustType {
+        let raw = unsafe { Schema_IndexBool(object.internal, field, index) };
+        raw != 0
+    }
+}
+
+impl<T: IndexedField> Sealed for Option<T> {}
+
+impl<T: IndexedField> SchemaField for Option<T> {
```

Comment:
Might not be important here specifically, but for future reference.. 

Important to note `Option<T>` in Rust != `option<t>` in schemalang. In schemalang `option` has reference semantics, not value semantics.

This would translate strictly to `Option<Box<T>>` but we likely will want to optimize for this (as we can detect cycles in the schema structure that would require reference semantics). 

---

Example: 

This is perfectly valid schema. 
```
type Bar {
	Foo field =  1;
}

type Foo {
	option<Bar> recursive = 1;
}
```

Generating Rust code 1:1 would result in: 

```rust
pub struct Bar {
	pub field: Foo,
}

pub struct Foo {
	pub recursive: Option<Bar>,
}
```

which will fail to compile unless you `Box` or otherwise indirect the inner type of `recursive`.

Reply:
Ah, that's a good catch! I'll add a test case for this in the example schema file. I'll probably start off by having `option<T>` always generate `Option<Box<T>>`, and we can implement an optimized version at some point that skips boxing when not necessary.